File: prep/implementation/patterns_reference_imeplementation/sliding_window/LongestSubstringWithSameLettersAfterReplacement.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(n) / O(1)
public class LongestSubstringWithSameLettersAfterReplacement {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();
      in.nextLine();

      while (tt-- > 0) {
        String s = in.next().trim();
        int k = in.nextInt();

        int n = s.length();
        Map<Character, Integer> freq = new HashMap<>();
        int maxLen = 0, maxFreq = 0, windowStart = 0;

        for (int windowEnd = 0; windowEnd < n; windowEnd++) {
          char c = s.charAt(windowEnd);
          freq.put(c, freq.getOrDefault(c, 0) + 1);
          maxFreq = Math.max(maxFreq, freq.get(c));

          while (windowEnd - windowStart + 1 - maxFreq > k) {
            char d = s.charAt(windowStart);
            freq.put(d, freq.get(d) - 1);
            windowStart++;
          }

          maxLen = Math.max(maxLen, windowEnd - windowStart + 1);
        }

        System.out.println(maxLen);
      }
    }
  }
}

File: prep/implementation/practice_iterations/sliding_window_3/longest_substring_with_same_letters_after_replacement.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

// O(n) / O(1)
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, k;
  string s;

  cin >> tt;
  while (tt--) {
    cin >> s >> k;

    unordered_map<char, int> freq;
    int max_len = 0, window_start = 0;
    int max_freq = 0;

    for (int window_end = 0; window_end < s.size(); window_end++) {
      char r = s[window_end];

      if (freq.find(r) == freq.end()) {
        freq[r] = 1;
      } else {
        freq[r]++;
      }

      max_freq = max(max_freq, freq[r]);

      while (window_end - window_start + 1 - max_freq > k) {
        char l = s[window_start];
        freq[l]--;
        window_start++;
      }

      max_len = max(max_len, window_end - window_start + 1);
    }

    cout << max_len << "\n";
  }

  return 0;
}